Add Earth Orbit Data Service

Clark199909 opened the pull request:
Earth Orbit Data

brollb commented:
Hi @Clark199909, I haven't looked into this much since I believe you are still working on this. Let me know when you want me to take a look at this PR :)

brollb commented on src/server/services/procedures/icesheets/icesheets.js:
```diff
@@ -0,0 +1,169 @@
+ */
+Icesheets.getsspecMapTime = function() {
+    return this._data3
+        .map(data => [data.specMap, data.LR04b]);
```
Since we return both specMap and LR04b, would it make more sense to reference something about this in the help comment? I am honestly not sure as I am not familiar with either of these concepts :/ 

Either way, my point is that we should make sure it is clear about the meaning of the returned data. We also should make sure that the order of the columns is intuitive. For example, it could make more sense to switch the columns. (Again, I am not familiar enough with the concepts to be able to say definitively.)

brollb commented on test/unit/server/services/procedures/icesheets.spec.js:
```diff
@@ -0,0 +1,12 @@
+describe('icesheets', function() {
+    const utils = require('../../../../assets/utils');
+
+    utils.verifyRPCInterfaces('EarthOrbit', [
```
This should be failing currently. Is it not? (`'EarthOrbit'` should be update to the IceSheets service name.)

brollb commented on src/server/services/procedures/earth-orbit/earth-orbit.js:
```diff
@@ -0,0 +1,164 @@
+EarthOrbit.get2004Longitude = function(startyear, endyear) {
+    return this._data1
+        .map(data => [data.year, Number(data.longitude)])
+        .filter(data => data[0] >= startyear && data[0] <= endyear);
```
Here is an example: https://github.com/NetsBlox/NetsBlox/blob/master/test/unit/server/services/procedures/cloud-variables.spec.js#L95-L98

brollb commented on src/server/services/procedures/earth-orbit/earth-orbit.js:
```diff
@@ -0,0 +1,164 @@
+    .forEach(line=>importDataPrec(line));
+
+/**
+ * Get longitude of perihelion from moving equinox by year.
```
Is there an easy link that we could add here to help explain the concepts?

brollb commented on src/server/services/procedures/earth-orbit/earth-orbit.js:
```diff
@@ -0,0 +1,164 @@
+lines1.reverse();
+lines1.splice(0,1);
+let lines2 = fs.readFileSync(path.join(__dirname, 'INSOLN.LA2004.BTL.250.ASC'), 'utf8').split('\n');
+lines1.concat(lines2).forEach(line=>importData2004(line));
```
Personally, I prefer to move this logic to a different file like in the `IceCoreData` service: https://github.com/NetsBlox/NetsBlox/tree/master/src/server/services/procedures/ice-core-data

This also makes it easier to write checks into the code like: https://github.com/NetsBlox/NetsBlox/blob/master/src/server/services/procedures/ice-core-data/data/wais-co2/index.js#L19-L20

Clark199909 commented on src/server/services/procedures/earth-orbit/earth-orbit.js:
```diff
@@ -0,0 +1,164 @@
+
+// converting time to year CE
+let timeConversion = function(time) {
+    return (2000 + 1000*time);
```
In earth orbit data, it is not using the BP present, and its present is the year 2000.

brollb commented on src/server/services/procedures/icesheets/icesheets.js:
```diff
@@ -0,0 +1,134 @@
+Icesheets._dataSedim = [];
+
+// Initialize startyear to -5000000 CE and endyear to 0 CE if no inputs are given.
+let checkInputYear = function(startyear, endyear) {
```
This should be removed in favor of default parameters https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters.

brollb commented on src/server/services/procedures/icesheets/icesheets.js:
```diff
@@ -0,0 +1,134 @@
+// reading and importing delta 18O data
+const importdelta18O = function(line) {
+    let [time, d180, error] = line.trim().split(/\s+/);
+    let adjustTime = timeConversion(time);
```
Input validation and casting should be performed here. It will be more performant and should help discover bugs. For example:
```javascript
error = Number(error);
assert(!isNaN(error));  // you will need "const assert = require('assert');" earlier, of course
```

brollb replied:
Validation for the d18O would be good, too

brollb commented on src/server/services/procedures/icesheets/icesheets.js:
```diff
@@ -0,0 +1,134 @@
+// reading and importing Sedimentation Rate data
+const importSedim = function(line) {
+    let [time, averagedSedRates, normalizedSedRates] = line.trim().split(/\s+/);
+    let adjustTime = timeConversion(time);
```
Same as above.

brollb commented on src/server/services/procedures/earth-orbit/earth-orbit.js:
```diff
@@ -0,0 +1,192 @@
+// reading and importing precession data
+const importDataPrec = function(line) {
+    let [year, value] = line.trim().split(/\s+/);
+    year = timeConversion(year);
```
Like the comment on the IceSheets, it would be great to do the parsing and validating here.

brollb commented on src/server/services/procedures/earth-orbit/earth-orbit.js:
```diff
@@ -0,0 +1,164 @@
+ */
+EarthOrbit.get2004Eccentricity = function(startyear, endyear) {
+    return this._data1
+        .map(data => [data.year, Number(data.eccentricity)])
```
On second thought, it would be a good idea to change this. I made a comment earlier.

Clark199909 commented:
Dear Brian,
I have made changes to the code and updated them. Can you review it again?

Best,
Clark

On Fri, Jul 3, 2020 at 10:12 AM Brian Broll <notifications@github.com>
wrote:

> *@brollb* requested changes on this pull request.
>
> Nice work, Clark. I added a few more comments and resolved the ones that
> you have addressed!
> ------------------------------
>
> In src/server/services/procedures/earth-orbit/earth-orbit.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449590419&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646830751&sdata=rKLsB4NU%2FqswWDbZR9k1k8sFnOwA0Z8CV4Ww0tm1i84%3D&reserved=0>
> :
>
> > +        .filter(data => data[0] >= startyear && data[0] <= endyear);
> +};
> +
> +/**
> + * Get obliquity by year. For more information about obliquity, please visit:
> + * https://climate.nasa.gov/news/2948/milankovitch-orbital-cycles-and-their-role-in-earths-climate/
> + *
> + * If a start and end year is provided, only measurements within the given range will be
> + * returned.
> + *
> + * @param {Number=} startyear Year to begin data at
> + * @param {Number=} endyear Year to begin data at
> + * @returns {Array} obliquity
> + */
> +EarthOrbit.get2004Obliquity = function(startyear, endyear) {
> +    [startyear,endyear] = checkInputYear(endyear,startyear);
>
> There is actually an easier way to do this :) Just update the function to
> be:
>
> EarthOrbit.get2004Obliquity = function(startyear=-Infinity, endyear=Infinity) {
>
> ------------------------------
>
> In src/server/services/procedures/earth-orbit/earth-orbit.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449595751&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646840748&sdata=1nYaHvI6enPrNaEDvQeInO0R%2Bg2pRPsBL19Tyg5BIMw%3D&reserved=0>
> :
>
> > +fs.readFileSync(path.join(__dirname, 'precession.txt'), 'utf8')
> +    .split('\n')
> +    .forEach(line=>importDataPrec(line));
> +
> +/**
> + * Get longitude of perihelion from moving equinox by year. For more information about this, please visit:
> + * https://www.physics.ncsu.edu/classes/astron/orbits.html
> + *
> + * If a start and end year is provided, only measurements within the given range will be
> + * returned.
> + *
> + * @param {Number=} startyear Year to begin data at
> + * @param {Number=} endyear Year to begin data at
> + * @returns {Array} longitude - longitude of perihelion from moving equinox
> + */
> +EarthOrbit.get2004Longitude = function(startyear, endyear) {
>
> Do you think it would be better to say "La2004" instead of just 2004? (I
> am not familiar enough with the domain to say definitively myself but I
> kept thinking this was referring to data from 2004 rather than a reference
> to a solution from a paper.)
>
> Another approach would be to specify this detail in the help message
> (maybe adding something like "... moving equinox by year using the solution
> from Laskar, et al., 2004").
> ------------------------------
>
> In src/server/services/procedures/icesheets/icesheets.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449597960&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646840748&sdata=A%2BVZjsw1tUxMkFfQmSbekTCR41N9R4qun%2F08i3uOiHs%3D&reserved=0>
> :
>
> > + * Access to NOAA Global Pliocene-Pleistocene Benthic d18O Stack.
> + *
> + * For more information, check out
> + * https://www.ncdc.noaa.gov/paleo-search/study/5847
> + *
> + * Original datasets are available at
> + * https://www1.ncdc.noaa.gov/pub/data/paleo/contributions_by_author/lisiecki2005/lisiecki2005.txt
> + * @alpha
> + * @service
> + * @category Science
> + * @category Climate
> + */
> +const fs = require('fs');
> +const path = require('path');
> +
> +const Icesheets = {};
>
> Ice sheets is two words so this should actually be "IceSheets". If you
> rename the containing directory (and js file) to "ice-sheets", it will
> capitalize it appropriately for the netsblox client, too.
> ------------------------------
>
> In src/server/services/procedures/icesheets/icesheets.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449598296&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646850737&sdata=L3HRyIdY12%2FhDQ%2FKKveKSmDdIU9GV34eIU8hAutZUeM%3D&reserved=0>
> :
>
> > + * @category Climate
> + */
> +const fs = require('fs');
> +const path = require('path');
> +
> +const Icesheets = {};
> +
> +// this contains delta 18O data, which is a ratio of two oxygen isotopes in deep ocean sediments
> +Icesheets._dataDelta18O = [];
> +
> +// this contains sedimentation rate data, which is a formation rate that is the amount of sediment
> +// deposited during a certain time span, normally expressed as a length per time (e.g., cm/yr).
> +Icesheets._dataSedim = [];
> +
> +// Initialize startyear to -5000000 CE and endyear to 0 CE if no inputs are given.
> +let checkInputYear = function(startyear, endyear) {
>
> This should be removed in favor of default parameters
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FJavaScript%2FReference%2FFunctions%2FDefault_parameters&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646850737&sdata=nOSP%2FRO%2BdolpNwMAv%2FMIKZQHaWIbMeIZ7u6AuaKOqBU%3D&reserved=0>
> .
> ------------------------------
>
> In src/server/services/procedures/icesheets/icesheets.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449599181&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646850737&sdata=lpmAuNmVUSX1dtwiiqb%2B6HdROvqoMuM5fWugRQhB0X4%3D&reserved=0>
> :
>
> > +    }
> +    if(endyear == null) {
> +        endyear = 0;
> +    }
> +    return [startyear, endyear];
> +};
> +
> +// converting time to year CE, note: the present is 1950 here because we are using the BP present
> +let timeConversion = function(time) {
> +    return (1950 - 1000*time);
> +};
> +
> +// reading and importing delta 18O data
> +const importdelta18O = function(line) {
> +    let [time, d180, error] = line.trim().split(/\s+/);
> +    let adjustTime = timeConversion(time);
>
> Input validation and casting should be performed here. It will be more
> performant and should help discover bugs. For example:
>
> error = Number(error);assert(!isNaN(error));  // you will need "const assert = require('assert');" earlier, of course
>
> ------------------------------
>
> In src/server/services/procedures/icesheets/icesheets.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449599305&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646860732&sdata=Qk7w4vRZwb6Mpc0rdi7uwR3WhFBVDq%2FH8tvaTkZDHIY%3D&reserved=0>
> :
>
> > +    }
> +    if(endyear == null) {
> +        endyear = 0;
> +    }
> +    return [startyear, endyear];
> +};
> +
> +// converting time to year CE, note: the present is 1950 here because we are using the BP present
> +let timeConversion = function(time) {
> +    return (1950 - 1000*time);
> +};
> +
> +// reading and importing delta 18O data
> +const importdelta18O = function(line) {
> +    let [time, d180, error] = line.trim().split(/\s+/);
> +    let adjustTime = timeConversion(time);
>
> Validation for the d18O would be good, too
> ------------------------------
>
> In src/server/services/procedures/icesheets/icesheets.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449599685&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646860732&sdata=kVtEhyviT7tP3YEwIKsS1cqX25bq%2FWA%2FiVJ73FPa38I%3D&reserved=0>
> :
>
> > +// converting time to year CE, note: the present is 1950 here because we are using the BP present
> +let timeConversion = function(time) {
> +    return (1950 - 1000*time);
> +};
> +
> +// reading and importing delta 18O data
> +const importdelta18O = function(line) {
> +    let [time, d180, error] = line.trim().split(/\s+/);
> +    let adjustTime = timeConversion(time);
> +    Icesheets._dataDelta18O.push({adjustTime, d180, error});
> +};
> +
> +// reading and importing Sedimentation Rate data
> +const importSedim = function(line) {
> +    let [time, averagedSedRates, normalizedSedRates] = line.trim().split(/\s+/);
> +    let adjustTime = timeConversion(time);
>
> Same as above.
> ------------------------------
>
> In src/server/services/procedures/earth-orbit/earth-orbit.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449602732&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646870729&sdata=Rx3JT4rTtnnOSTGCeyG3DoIo9EqlR%2FW%2BLLFRmWGQwvo%3D&reserved=0>
> :
>
> > +    obliquity = parseDecimalValue(obliquity);
> +    longitude = parseDecimalValue(longitude);
> +    EarthOrbit._dataEccOblLongi.push({year, eccentricity, obliquity, longitude});
> +};
> +
> +// reading and importing insolation data
> +const importDataInsol = function(line) {
> +    let [year, value] = line.trim().split(/\s+/);
> +    year = timeConversion(year);
> +    EarthOrbit._dataInsolation.push({year, value});
> +};
> +
> +// reading and importing precession data
> +const importDataPrec = function(line) {
> +    let [year, value] = line.trim().split(/\s+/);
> +    year = timeConversion(year);
>
> Like the comment on the IceSheets, it would be great to do the parsing and
> validating here.
> ------------------------------
>
> In src/server/services/procedures/earth-orbit/earth-orbit.js
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23discussion_r449602956&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646870729&sdata=KkISCKbBy%2BaU8f5gxEQ9PcbZOS1%2Fv0Nb9HdBpoLrgvk%3D&reserved=0>
> :
>
> > +        .filter(data => data[0] >= startyear && data[0] <= endyear);
> +};
> +
> +/**
> + * Get eccentricity by year.
> + *
> + * If a start and end year is provided, only measurements within the given range will be
> + * returned.
> + *
> + * @param {Number=} startyear Year to begin data at
> + * @param {Number=} endyear Year to begin data at
> + * @returns {Array} eccentricity
> + */
> +EarthOrbit.get2004Eccentricity = function(startyear, endyear) {
> +    return this._data1
> +        .map(data => [data.year, Number(data.eccentricity)])
>
> On second thought, it would be a good idea to change this. I made a
> comment earlier.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FNetsBlox%2FNetsBlox%2Fpull%2F2804%23pullrequestreview-442422073&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646880722&sdata=4aEJvqk%2BVikfQqn%2B66c7i86Kx8cPRScBDRfrqi%2FrutE%3D&reserved=0>,
> or unsubscribe
> <https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAM7NEYPRNURLCXPZE2MQGKTRZXRNXANCNFSM4N4TEWGA&data=02%7C01%7Czhengkai.xie%40mail-service-3-mx.vanderbilt.edu%7C2056bb884e034982469c08d81f5b277d%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C0%7C637293823646880722&sdata=UaN7OxQBZwXllVOLfpQAVpO9LKZf1dvNkgFcIpNfbIw%3D&reserved=0>
> .
>

brollb commented on src/server/services/procedures/paleocean-oxygen-isotopes/paleocean-oxygen-isotopes.js:
```diff
@@ -0,0 +1,120 @@
+PaleoceanOxygenIsotopes.getAverageSedimentationRates = function(startyear = -Infinity, endyear = Infinity) {
+    return this._dataSedim
+        .map(data => [data.adjustTime, data.averagedSedRates])
+        .filter(data => data[0] >= startyear && data[0] <= endyear);
```
This actually would be more legible (and slightly more performant) if you first filter than map:
```javascript
return this._dataSedim
    .filter(data => data.adjustTime >= startyear && data.adjustTime <= endyear)
    .map(data => [data.adjustTime, data.averagedSedRates]);
```

brollb replied:
Also, maybe just recording the time as `year` (or even `yearCE`) rather than `adjustTime` might be a little more clear since it doesn't really matter that it had to be adjusted for the downstream usage (like here).

brollb commented on test/unit/server/services/procedures/earth-orbit.spec.js:
```diff
@@ -0,0 +1,11 @@
+    const utils = require('../../../../assets/utils');
+
+    utils.verifyRPCInterfaces('EarthOrbit', [
+        ['get2004Longitude', ['startyear', 'endyear']],
+        ['get2004Obliquity', ['startyear', 'endyear']],
+        ['get2004Eccentricity', ['startyear', 'endyear']],
+        ['get2004Insolation', ['startyear', 'endyear']],
+        ['get2004Precession', ['startyear', 'endyear']]
```
These RPC names need to be updated.

brollb commented on test/unit/server/services/procedures/icesheets.spec.js:
```diff
@@ -0,0 +1,10 @@
+describe('Icesheets', function() {
```
This file needs to be removed

brollb commented:
Nice work, @Clark199909! LGTM
